File: packages/transform/src/features/llm/LlmCreateParseTransformer.ts
import {
  LlmMetadataFactory,
  LlmParseProgrammer,
  MetadataCollection,
  MetadataFactory,
  MetadataSchema,
} from "@typia/core";
import { ILlmSchema, ValidationPipe } from "@typia/interface";
import ts from "typescript";

import { ITransformProps } from "../../ITransformProps";
import { TransformerError } from "../../TransformerError";

export namespace LlmCreateParseTransformer {
  export const transform = (props: ITransformProps): ts.Expression => {
    // GET GENERIC ARGUMENT
    if (!props.expression.typeArguments?.length)
      throw new TransformerError({
        code: "typia.llm.createParse",
        message: "no generic argument.",
      });

    const top: ts.Node = props.expression.typeArguments[0]!;
    if (ts.isTypeNode(top) === false) return props.expression;

    // GET TYPE AND CONFIG
    const config: Partial<ILlmSchema.IConfig> = LlmMetadataFactory.getConfig({
      context: props.context,
      method: "createParse",
      node: props.expression.typeArguments[1],
    }) as Partial<ILlmSchema.IConfig>;

    const type: ts.Type = props.context.checker.getTypeFromTypeNode(top);

    if (type.isTypeParameter())
      throw new TransformerError({
        code: "typia.llm.createParse",
        message: "non-specified generic argument.",
      });

    // VALIDATE TYPE
    const result: ValidationPipe<MetadataSchema, MetadataFactory.IError> =
      MetadataFactory.analyze({
        checker: props.context.checker,
        transformer: props.context.transformer,
        options: {
          absorb: true,
          escape: true,
          constant: true,
          validate: (metadata, explore) =>
            LlmParseProgrammer.validate({
              config,
              metadata,
              explore,
            }),
        },
        components: new MetadataCollection({
          replace: MetadataCollection.replace,
        }),
        type,
      });
    if (result.success === false)
      throw TransformerError.from({
        code: "typia.llm.createParse",
        errors: result.errors,
      });

    // GENERATE CODE (factory returns the function directly)
    return LlmParseProgrammer.write({
      context: props.context,
      modulo: props.modulo,
      type,
      name: (top as ts.TypeNode).getFullText().trim(),
      config,
    });
  };
}

File: packages/typia/native/transform/features/llm/LlmCreateParseTransformer.go
package llm

import (
  shimast "github.com/microsoft/typescript-go/shim/ast"
  nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
  nativellmprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers/llm"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
  nativetransform "github.com/samchon/typia/packages/typia/native/transform/internal"
)

type llmCreateParseTransformerNamespace struct{}

var LlmCreateParseTransformer = llmCreateParseTransformerNamespace{}

func (llmCreateParseTransformerNamespace) Transform(props nativetransform.ITransformProps) *shimast.Node {
  top, typ, ok := llmTransformer_type_argument(props, "typia.llm.createParse")
  if ok == false {
    return props.Expression.AsNode()
  }
  config := llmTransformer_config(props, "createParse")
  if typ != nil && typ.IsTypeParameter() {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{Code: "typia.llm.createParse", Message: "non-specified generic argument."}))
  }
  analyze := func(validate bool) *schemametadata.MetadataSchema {
    return llmTransformer_analyze(llmTransformer_analyzeProps{
      Context: props.Context,
      Type:    typ,
      Code:    "typia.llm.createParse",
      Absorb:  validate,
      Validate: func(next struct {
        Metadata *schemametadata.MetadataSchema
        Explore  nativefactories.MetadataFactory_IExplore
        Top      *schemametadata.MetadataSchema
      }) []string {
        if validate == false {
          return nil
        }
        return nativellmprogrammers.LlmParseProgrammer.Validate(struct {
          Config   map[string]any
          Metadata *schemametadata.MetadataSchema
          Explore  nativefactories.MetadataFactory_IExplore
        }{Config: config, Metadata: next.Metadata, Explore: next.Explore})
      },
    })
  }
  analyze(true)
  return nativellmprogrammers.LlmParseProgrammer.Write(nativellmprogrammers.LlmParseProgrammer_IWriteProps{
    Context:  props.Context,
    Modulo:   props.Modulo,
    Metadata: analyze(false),
    Name:     llmTransformer_type_name(top),
    Config:   config,
  })
}
